File: CleanCity/app/src/main/java/com/example/cleancity/api/RecorridoConexion.java
package com.example.cleancity.api;

import android.util.Log;

import com.example.cleancity.interfaces.ICallBackFunciono;
import com.example.cleancity.interfaces.ICallBackRecorrido;
import com.example.cleancity.interfaces.ICallBackUsuario;
import com.example.cleancity.modelos.RecorridoModelo;
import com.example.cleancity.modelos.UsuarioModelo;

import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Response;

public class RecorridoConexion extends ConexionBase{
    public void seleccionar(ICallBackRecorrido callback){
        solicitar("https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/seleccionar.php", new ArrayList(), new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.d("Respuesta", e.getMessage());
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                String respuesta = response.body().string();
                try{
                    JSONArray lista = new JSONArray(respuesta);
                    List<RecorridoModelo> listaRetorno = new ArrayList<>();
                    for (int i = 0; i < lista.length(); i++){
                        RecorridoModelo rec = new RecorridoModelo(lista.getJSONObject(i));
                        listaRetorno.add(rec);
                    }
                    callback.retornar(listaRetorno);
                }catch (Exception e){
                    e.printStackTrace();
                }
            }
        });
    }

    public void buscar(RecorridoModelo rec, ICallBackRecorrido callback){
        solicitar("https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/buscar.php", rec.busqueda(), new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.d("Respuesta", e.getMessage());
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                String respuesta = response.body().string();
                try{
                    JSONArray lista = new JSONArray(respuesta);

                    RecorridoModelo recorrido = new RecorridoModelo(lista.getJSONObject(0));

                    callback.retorno(recorrido);
                }catch (Exception e){
                    e.printStackTrace();
                }
            }
        });
    }

    public void insertar(RecorridoModelo rec, ICallBackFunciono callback){
        solicitar("https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/insertar.php", rec.parametros(), new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.d("Respuesta", e.getMessage());
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                try{
                    String funciono = response.body().string();
                    callback.retornar(Boolean.parseBoolean(funciono));
                } catch (Exception e){}
            }
        });
    }

    public void editar(RecorridoModelo rec, ICallBackFunciono callback){
        solicitar("https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/editar.php", rec.parametros(), new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.d("Respuesta", e.getMessage());
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                try{
                    String funciono = response.body().string();
                    callback.retornar(Boolean.parseBoolean(funciono));
                } catch (Exception e){}
            }
        });
    }

    public void eliminar(RecorridoModelo rec, ICallBackFunciono callback){
        solicitar("https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/eliminar.php", rec.busqueda(), new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.d("Respuesta", e.getMessage());
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                try{
                    String funciono = response.body().string();
                    callback.retornar(Boolean.parseBoolean(funciono));
                } catch (Exception e){

                }
            }
        });
    }
}

File: CleanCityClient/app/src/main/java/com/example/cleancityclient/api/RecorridoConexion.kt
package com.example.cleancityclient.api

import android.util.Log
import com.example.cleancityclient.interfaces.ICallBackFunciono
import com.example.cleancityclient.interfaces.ICallBackRecorrido
import com.example.cleancityclient.modelos.RecorridoModelo
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import org.json.JSONArray
import java.io.IOException
import java.lang.Boolean.parseBoolean
import java.util.*

class RecorridoConexion: ConexionBase() {
    fun seleccionar(callback: ICallBackRecorrido) {
        solicitar(
            "https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/seleccionar.php",
            ArrayList<Parametro>(),
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.d("RESPONSE", e.message!!)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    val respuesta = response.body!!.string()
                    try {
                        val lista = JSONArray(respuesta)
                        val listaRetorno: MutableList<RecorridoModelo> =
                            ArrayList<RecorridoModelo>()
                        for (i in 0 until lista.length()) {
                            val rec = RecorridoModelo(lista.getJSONObject(i))
                            listaRetorno.add(rec)
                        }
                        callback.retornar(listaRetorno)
                    } catch (e: Exception) {
                        Log.d("RESPONSE", e.message!!)
                        e.printStackTrace()
                    }
                }
            })
    }

    fun buscar(rec: RecorridoModelo, callback: ICallBackRecorrido) {
        solicitar(
            "https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/buscar.php",
            rec.busqueda(),
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.d("RESPONSE", e.message!!)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    val respuesta = response.body!!.string()
                    try {
                        val lista = JSONArray(respuesta)
                        val recorrido = RecorridoModelo(lista.getJSONObject(0))
                        callback.retorno(recorrido)
                    } catch (e: Exception) {
                        Log.d("RESPONSE", e.message!!)
                        e.printStackTrace()
                    }
                }
            })
    }

    fun insertar(rec: RecorridoModelo, callback: ICallBackFunciono) {
        solicitar(
            "https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/insertar.php",
            rec.parametros(),
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.d("RESPONSE", e.message!!)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        val funciono = response.body!!.string()
                        callback.retornar(parseBoolean(funciono))
                    } catch (e: Exception) {
                        Log.d("RESPONSE", e.message!!)
                        e.printStackTrace()
                    }
                }
            })
    }

    fun editar(rec: RecorridoModelo, callback: ICallBackFunciono) {
        solicitar(
            "https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/editar.php",
            rec.parametros(),
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.d("RESPONSE", e.message!!)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        val funciono = response.body!!.string()
                        callback.retornar(parseBoolean(funciono))
                    } catch (e: Exception) {
                        Log.d("RESPONSE", e.message!!)
                        e.printStackTrace()
                    }
                }
            })
    }

    fun eliminar(rec: RecorridoModelo, callback: ICallBackFunciono) {
        solicitar(
            "https://cleancitysp.000webhostapp.com/api/solicitudes/recorrido/eliminar.php",
            rec.busqueda(),
            object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.d("RESPONSE", e.message!!)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        val funciono = response.body!!.string()
                        callback.retornar(parseBoolean(funciono))
                    } catch (e: Exception) {
                        Log.d("RESPONSE", e.message!!)
                        e.printStackTrace()
                    }
                }
            })
    }
}